### DISCUSSION TITLE
Strategies for debugging?

### DISCUSSION THREAD

**Discussion by devo-wm** [silviogutierrez/reactivated#229] [Q&A]
In my react projects I tend to used `console.log` and `debugger` to inspect and pause scripts while running, but since these are being compiled server side I'm not seeing these any in the browser's console.

Since I use VS Code, I was wondering: does anyone know of a way to debug the tsx/jsx templates in a similar manner as describe here for the django template debugging? https://code.visualstudio.com/docs/python/tutorial-django#_use-the-debugger-with-page-templates

**Comment by silviogutierrez** [comment]
@devo-wm : I'm curious how to do this as well. For the record, `console.log` should absolutely work and I use it all the time. While I don't see the server side logs, the client side hydrates and I do see information.

One option for `console.log` to work server side is simply to enable pass through std output in the subprocess that is opened:

https://github.com/silviogutierrez/reactivated/blob/main/reactivated/processes.py

Where the renderer is started, notice we initially capture stdout to get the socket, but then no longer bubble stuff through. Unfortunately, I'm not familiar enough with suprocess to make this happen. But you can see the `tsc` subprocess does this just fine (since we don't need to process its output). Try starting the renderer just like `tsc` and set `os.environ["REACTIVATED_RENDERER"] = renderer_process_port` manually instead of capturing output. I think output should bubble through to `runserver`

As for using `debugger`, theoretically whatever people in a pure NodeJS SSR context do should work here. I'm not sure what they do though.

**Comment by rix1** [comment]
Hi @silviogutierrez!
This project looks amazing - great work! 🙌 I'm playing around with a demo app to evaluate if this could be a viable solution to a large existing application that is in dire need of better frontend tools, but without compromising the power of Django 🙏 

As I'm working through, I tried adding [Django debug toolbar](https://django-debug-toolbar.readthedocs.io/en/latest/), but it seems that the `reactivated.backend.JSX` template processor doesn't allow middleware injecting things to the rendered page(?). Maybe I'm on the wrong path here, but have anyone else added Django Debug Toolbar (or something similar) to their application? 

Note: This isn't strictly related to the first question asked - let me know if I should open a new discussion!

**Reply by silviogutierrez** [reply]
@rix1 : thanks for your kind words. So in most cases, you're not actually using the template engine. Rather, you're calling `MyTemplate().render(request)`, see the sample views included with the project.

With that said, I just tried enabling the debug toolbar and it worked fine, see here:

<img width="2560" alt="image" src="https://user-images.githubusercontent.com/92824/229377340-68133891-ee3d-4da0-8709-452c1ca14762.png">

And here is a git diff of a clean project:

```diff
diff --git a/server/settings/common.py b/server/settings/common.py
index ded7159..92520e3 100644
--- a/server/settings/common.py
+++ b/server/settings/common.py
@@ -37,6 +37,7 @@ ALLOWED_HOSTS = ["*"]
 # Application definition

 INSTALLED_APPS = [
+    "debug_toolbar",
     "server.example",
     "django.contrib.admin",
     "django.contrib.auth",
@@ -48,6 +49,7 @@ INSTALLED_APPS = [
 ]

 MIDDLEWARE = [
+    "debug_toolbar.middleware.DebugToolbarMiddleware",
     "django.middleware.security.SecurityMiddleware",
     "django.contrib.sessions.middleware.SessionMiddleware",
     "django.middleware.common.CommonMiddleware",
@@ -143,3 +145,7 @@ DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
 STATICFILES_DIRS = (BASE_DIR / "static/",)

 STATIC_ROOT = BASE_DIR / "collected/"
+
+INTERNAL_IPS = [
+    "127.0.0.1",
+]
diff --git a/server/urls.py b/server/urls.py
index be12f12..b904495 100644
--- a/server/urls.py
+++ b/server/urls.py
@@ -3,5 +3,6 @@ from django.urls import include, path

 urlpatterns = [
     path("admin/", admin.site.urls),
+    path('__debug__/', include('debug_toolbar.urls')),
     path("", include("server.example.urls")),
 ]
```

**Reply by rix1** [reply]
Ah! Thanks 🙌 I was missing the `INTERNAL_IPS` config, but now it works! Are you looking for maintainers? Do you have a roadmap, or is this mostly covered by the RFCs? (This project deserves way more attention IMO)
